teps/0140-producing-results-in-matrix.md

```diff
@@ -0,0 +1,512 @@
+
+### Non-Goals
+
+The following are out of scope for this TEP:
+
+1. Support a matrixed `PipelineTask` that produces `Results` of type `Array` or `Object` that are then consumed by another `PipelineTask`.
+    - A matrixed `PipelineTask` can support `Results` of any type as long as they aren’t consumed by another `PipelineTask`. However, only matrixed `PipelineTask` that produces `Results` of type `String` can be consumed by another `PipelineTask`.
+2. Support consuming a specific instance or combination(s) of `Results` produced by a fanned out `PipelineTask`.
+    - At this time, we propose only supporting whole `Array` `Result` replacements from a Matrixed `PipelineTask`
+
+### Background
+
+Consuming `Results` from previous `TaskRuns` or `Runs` in a `Matrix`, which would dynamically generate `TaskRuns` from the fanned out `Results`, is currently supported for string `Results`, array `Results`, and object `Results`. Note that the underlying `Results` type in each must be string.
+
+#### String Results
+String `Results` is a stable feature and is referred to as `$(tasks.<pipelinetask-name>.results.<result-name>)` and can be passed into the `Matrix` from previous `TaskRuns`.
+```yaml
+tasks:
+...
+- name: task-1
+  taskRef:
+    name: task-1
+  matrix:
+  - name: values
+    value:
+    - (tasks.task-1.results.foo) # string
+```
+#### Array Results
+Array `Results` is a beta feature and is referred to as `$(tasks.<pipelinetask-name>.results.<result-name>[*])`. String replacements from arrays are supported through Array indexing and are referred to as `$(tasks.<pipelinetask-name>.results.<result-name>[i])`  where i is the index. Array `Results` from previous `TaskRuns` can be passed into the Matrix:
+```yaml
+tasks:
+...
+- name: task-2
+  taskRef:
+    name: task-2
+  matrix:
+  - name: values
+    value: (tasks.task-4.results.bar[*]) # array
+```
+
+#### Object Results
+Object `Results` is a beta feature and is referred to as `$(tasks.<pipelinetask-name>.results.<result-name>.[*])`. String replacements from objects are supported and are referred to as $(tasks.<pipelinetask-name>.results.<result-name>.key)  where key is the object key. Strings from Object `Results` from previous `TaskRuns` can be passed into the `Matrix`:
+tasks:
+
+```yaml
+...
+- name: task-3
+  taskRef:
+    name: task-3
+  matrix:
+  - name: values
+    value: (tasks.task-4.results.rad.key) # string replacement from object result
+```
+
+
+### Requirements
+
+1. A `matrix` of `Parameters` can only produce `Results` of type `string` if the `Results` are being consumed by another `PipelineTask`.
+2. If the `Results` from a `Matrixed` `PipelineTask` are *not* being referenced by another `PipelineTask`, the `Results` must have an underlying type `string`, similar to how `params` have an underlying type `string`.
+3. A `PipelineTask` that consumes a `Result` produced by a `Matrix` must consume the entire aggregated `Array` of `Results` produced.
+
+### Use Cases
```

Comment:
One note : all use cases are targeting `CI`, do we have any other use cases (CD, mlops, generic workflows, …) ?

Reply:
@pritidesai will add additional examples.